### ISSUE TITLE
uilist shakes violently: AIM > examine(action menu) on small screens

### ISSUE THREAD

**Issue by Brambor** [CleverRaven/Cataclysm-DDA#78783]
### Describe the bug

It shakes!

https://github.com/user-attachments/assets/7b983dfe-62b3-431f-a1db-288e0364d821

### Attach save file

N/A

### Steps to reproduce

1. Make screen very small
2. `/` open AIM
3. `e`xamine item in your inventory
4. Observe: uilist pops up that shakes violently

### Expected behavior

Uilist does not shake. Still scrollable.

### Versions and configuration

- OS: Windows
    - OS Version: 10.0.19045.5247 (22H2)
- Game Version: 0.G-16037-g3e310fe700 [64-bit] - master was 0664114d67
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Portal Storms Ignore NPCs [personal_portal_storms],
    Slowdown Fungal Growth [no_fungal_growth]
]

### Additional context

Debug menu uilists are fine:
![image](https://github.com/user-attachments/assets/b7a47172-4ab8-445e-95bb-76f92399ce19)

**Comment by Kilvoctu** [comment]
I can confirm this happens very frequently, as one of the platforms I play on is a 640x480 screen handheld gaming device (currently 80x32 terminal size). It only seems to affect that specific type of popup

**Comment by CoroNaut** [comment]
#76215 Similar issue, don't know if its a duplicate.

**Comment by Procyonae** [comment]
Still present on 0.I branch, can even happen on 1920x1080 for large menus like the debug map editor trap/terrain/furniture placement

**Comment by RenechCDDA** [comment]
Does this still occur outside of the debug menu? If it only happens in the debug menu it's not a blocker.

**Comment by Procyonae** [comment]
https://github.com/user-attachments/assets/67898c00-2139-49f0-90e7-2525457706e7

Yup, I were just pointing out it's not exclusively small screens and appears to be worse relative to the size of the list compared to the size of the screen

Edit: Closing the older issue bc this one correctly ids that this is an all uilist problem and has more discussion

**Comment by RenechCDDA** [comment]
@db48x Do you have any wisdom to share here? I've spent a few hours narrowing it down and as far as I can tell, something is wrong with how imgui's clipper calculates the starting position. Even without any new input to ImGui::SetScrollHereY() it continually changes the scrolling position between frames in a manner I've not been able to figure out.

<img width="1458" height="899" alt="Image" src="https://github.com/user-attachments/assets/c44b05c2-b43f-48a7-affa-33e8ed0b3ffa" />

Image 1:

<img width="411" height="484" alt="Image" src="https://github.com/user-attachments/assets/6ce92ce7-03ad-4a93-8f36-11e28027a78d" />

Image 2:

<img width="337" height="518" alt="Image" src="https://github.com/user-attachments/assets/53f739d1-7964-4546-93c2-4702fe1e7e23" />

**Comment by db48x** [comment]
Weird. I thought it was just the debug menus too!

Are you sure the uilist isn’t calling `ImGui::SetScrollHereY();` anywhere in there?

**Comment by RenechCDDA** [comment]
The only place the uilist calls it is here: https://github.com/CleverRaven/Cataclysm-DDA/blob/218dc92f437dd8d510dab13362583656b791c8f9/src/uilist.cpp#L148-L151

This simply doesn't get reached at all outside of when we press up/down. (If you see the debug message "Scroll WAS..." and "Scroll AFTER IS NOW..." those were positioned immediately before and after the ImGui::SetScrollHereY() call, to see if we were setting it incorrectly. We're not.)

I also breakpointed ImGui::SetScrollHereY(), ImGui::SetScrollY(), and ImGui::SetScrollPosY() just to be sure. None of them ever get called while it's bouncing around. The clipper is simply clipping incorrectly in between frames.

**Comment by db48x** [comment]
Well, usually when something like this happens it’s because we’re holding it wrong, but sometimes it is actually a bug in ImGui. If we had a minimized test case then a bug report would get quick attention and we would soon learn what I did wrong. Minimizing the code shouldn’t be that difficult; just copy the body of `uilist::draw_controls` into another file, then put some code to start, position, and size a window around it. Finally supply it with some static data to give it the same content as one of our menus. You’ll know that you’re done when you can paste it into one of the examples and reproduce the bug. I can do it on Monday if you haven’t.

**Comment by RenechCDDA** [comment]
> I can do it on Monday if you haven’t.

That would be really appreciated.

**Comment by db48x** [comment]
Hmm. My minimized testcase doesn’t reproduce the problem. Are you pressing any keys or hovering the mouse over the window or something?

**Comment by RenechCDDA** [comment]
Nope. Completely hands off the keyboard, you can even disable mouse inputs.. Once any input has been given it just starts clipping back and forth.

**Comment by RenechCDDA** [comment]
I think it's worth noting that if you set the scrollbar to be always visible via flag, it continually changes size. Even before any inputs are applied in the menu.

**Comment by Brambor** [comment]
I am debugging it now. ~I think the issue is with:~

https://github.com/CleverRaven/Cataclysm-DDA/blob/7bd751cbd8042535dd9e0ae1d4eb50f70ed265be/src/uilist.cpp#L197-L199

Nevermind.

~I will test it and stuff.~ Heads up, so that our work doesn't duplicate.

**Comment by Brambor** [comment]
> Still present on 0.I branch, can even happen on 1920x1080 for large menus like the debug map editor trap/terrain/furniture placement

In edit fields, search for `water`, which has only one result. The scroll bar still sometimes appears for a brief moment multiple times with different sizes. Sometimes it moves the single item out of view! I know we focus on what the player can see, not debug. But the issue is likely related, so it's just better at demonstrating it.

**Comment by Brambor** [comment]
I made a minimal reproducible example: With this patch, on game start, a uilist will open. If you go down a bit in that list, it will start shaking violently:

```patch
diff --git a/src/main_menu.cpp b/src/main_menu.cpp
index c970251f65..2cbdc7c8b2 100644
--- a/src/main_menu.cpp
+++ b/src/main_menu.cpp
@@ -577,6 +577,14 @@ void main_menu::load_char_templates()
 
 bool main_menu::opening_screen()
 {
+    uilist fmenu;
+    for( int i = 0; i < 1000; i++ ) {
+        fmenu.addentry( std::to_string( i ) );
+    }
+    while( true ) {
+        // with false, it flickers
+        fmenu.query( false );
+    }
     // set holiday based on local system time
     current_holiday = get_holiday_from_time();

```

**Comment by Brambor** [comment]
That is it from me. I will no longer look into this. I wish the next guy/gal good luck!

**Comment by Brambor** [comment]
Nevermind. I lied. I have an idea. Just don't use `query(false)`, lol. Will try tomorrow.

Edit: I couldn't sleep on it. And made a PR.
